### PR TITLE
Move geocoding to data service

### PR DESCRIPTION
**Pull request by attwad**
For #1043

Very sorry about the size of the PR but I haven't found an easy way to split that up given how pressing this is.

Only did minor adjustments to unit test, mostly just moved them over to their new package and didn't touch the cypress tests at all which makes me fairly confident this is working :)

### CODE REVIEW

**Comment by attwad** [comment]
(Did some last minute change to upsert() after I had run all the integration tests, will fix whatever broke tomorrow.)

**Comment by axmb** [comment]
> For #1043
> 
> Very sorry about the size of the PR but I haven't found an easy way to split that up given how pressing this is.
> 
> Only did minor adjustments to unit test, mostly just moved them over to their new package and didn't touch the cypress tests at all which makes me fairly confident this is working :)

No worries at all on the size, totally understood. I've reviewed everything aside from the data service case controller/main app, but stopped at this comment: https://github.com/globaldothealth/list/pull/1054#discussion_r489586599.

As mentioned all good on the size, but from a record-keeping perspective I think cutting that change sepearately would be really nice, if it's not a PITA.

**Comment by attwad** [comment]
> > For #1043
> > Very sorry about the size of the PR but I haven't found an easy way to split that up given how pressing this is.
> > Only did minor adjustments to unit test, mostly just moved them over to their new package and didn't touch the cypress tests at all which makes me fairly confident this is working :)
> 
> No worries at all on the size, totally understood. I've reviewed everything aside from the data service case controller/main app, but stopped at this comment: [#1054 (comment)](https://github.com/globaldothealth/list/pull/1054#discussion_r489586599).
> 
> As mentioned all good on the size, but from a record-keeping perspective I think cutting that change sepearately would be really nice, if it's not a PITA.

I will look at that tomorrow but basically because I need to access `this.geocoders` it has to be a class as it was in the curator service, perhaps hiding whitespace helps? I haven't tried it much, nor the split view tbh
image

**Comment by axmb** [comment]
> > > For #1043
> > > Very sorry about the size of the PR but I haven't found an easy way to split that up given how pressing this is.
> > > Only did minor adjustments to unit test, mostly just moved them over to their new package and didn't touch the cypress tests at all which makes me fairly confident this is working :)
> > 
> > 
> > No worries at all on the size, totally understood. I've reviewed everything aside from the data service case controller/main app, but stopped at this comment: [#1054 (comment)](https://github.com/globaldothealth/list/pull/1054#discussion_r489586599).
> > As mentioned all good on the size, but from a record-keeping perspective I think cutting that change sepearately would be really nice, if it's not a PITA.
> 
> I will look at that tomorrow but basically because I need to access `this.geocoders` it has to be a class as it was in the curator service, perhaps hiding whitespace helps? I haven't tried it much, nor the split view tbh

Right, I meant making it a class in a separate PR (on which this PR would be based). But you've done one better -- hiding whitespace is perfect. :tada:

File: data-serving/data-service/src/controllers/case.ts [hunk 1]
```diff
@@ -6,74 +6,552 @@ import parseSearchQuery, { ParsingError } from '../util/search';
...
-export const download = async (req: Request, res: Response): Promise<void> => {
-    let cases: any;
-    try {
-        if (req.body.query) {
-            cases = await casesMatchingSearchQuery({
-                searchQuery: req.body.query as string,
-                count: false,
+type BatchValidationErrors = { index: number; message: string }[];
+
+export class CasesController {
```
[the remark this answers is not in the document]
Comment (axmb) [reply data-serving/data-service/src/controllers/case.ts:15]: (Big ups to Tim for enlightening me as to the `hide whitespace` option.)

File: data-serving/data-service/src/controllers/case.ts [hunk 2]
```diff
@@ -6,74 +6,552 @@ import parseSearchQuery, { ParsingError } from '../util/search';
...
+        if (!c) {
+            res.status(404).send({
+                message: `Case with ID ${req.params.id} not found.`,
+            });
+            return;
+        }
+        res.status(204).end();
+    };
+
+    /**
```
Comment (axmb) [inline data-serving/data-service/src/controllers/case.ts:507]: Believe here down is copy-pasta? If there's a changed section I missed, LMK.

**Review by axmb — APPROVED** [review]

**Comment by attwad** [comment]
FYI to fix the integration test I had to move the batch geocode method to a middleware that ran before other middlewares in batchUpsert otherwise the preprocessor that was removing unchanged cases wasn't working as intended (location was different because not present yet, your awesome cypress tests for bulk upload caught that issue). This is cleaner and we can consider moving batchValidate as a middleware as well later on.

File: aws/data.yaml [hunk 3]
```diff
@@ -43,6 +43,11 @@ spec:
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.labels['environment']
+            - name: MAPBOX_TOKEN
+              valueFrom:
+                secretKeyRef:
+                  name: curator-dev-bch28c9gkm
```
[the remark this answers is not in the document]
Comment (attwad) [reply aws/data.yaml:49]: yeah I have mixed feelings about that one, there is no strong separation of secrets between services (no acl checks are done to see if they have access to it or not) so imo no big deal if that's named after a secret that was originally read from the curator service. We could remove that (data-|curator-) prefix altogether as a better solution .

**Pull request merged**